Sources/StreamChat/Controllers/DatabaseObserver/StateLayerListDatabaseObserver.swift

```diff
@@ -0,0 +1,130 @@
+    private(set) var resultsDelegate: FetchedResultsDelegate?
+    private let queue = DispatchQueue(label: "io.getstream.streamchat.statelayerdatabaseobserver")
+
+    let itemCreator: (DTO) throws -> Item
+    let sorting: [SortValue<Item>]
+    let request: NSFetchRequest<DTO>
+    let context: NSManagedObjectContext
+    
+    init(
+        context: NSManagedObjectContext,
+        fetchRequest: NSFetchRequest<DTO>,
+        itemCreator: @escaping (DTO) throws -> Item,
+        sorting: [SortValue<Item>] = [],
+        fetchedResultsControllerType: NSFetchedResultsController<DTO>.Type = NSFetchedResultsController<DTO>.self
+    ) {
+        self.context = context
+        request = fetchRequest
+        self.itemCreator = itemCreator
+        self.sorting = sorting
+        frc = fetchedResultsControllerType.init(
+            fetchRequest: request,
+            managedObjectContext: context,
+            sectionNameKeyPath: nil,
+            cacheName: nil
+        )
+    }
+    
+    convenience init(
+        databaseContainer: DatabaseContainer,
+        fetchRequest: NSFetchRequest<DTO>,
+        itemCreator: @escaping (DTO) throws -> Item,
+        sorting: [SortValue<Item>]
+    ) {
+        // We must use the writableContext since state layer needs to react to the change immediately.
+        // Otherwise async functions updating the CoreData store can return before we have updated the respective state object.
+        self.init(
+            context: databaseContainer.writableContext,
+            fetchRequest: fetchRequest,
+            itemCreator: itemCreator,
+            sorting: sorting
+        )
+    }
+    
+    var items: StreamCollection<Item> {
+        queue.sync { _items }
+    }
+
+    private var _items = StreamCollection<Item>([])
+    
+    func startObserving(didChange: ((StreamCollection<Item>) async -> Void)? = nil) throws {
+        resultsDelegate = FetchedResultsDelegate(onDidChange: { [weak self] in
+            guard let self else { return }
+            // Runs on the NSManagedObjectContext's queue
+            let collection = Self.makeCollection(
+                frc: self.frc,
+                context: self.context,
+                itemCreator: self.itemCreator,
+                sorting: self.sorting
+            )
+            queue.sync {
+                self._items = collection
+            }
+            Task { await didChange?(collection) }
+        })
+        frc.delegate = resultsDelegate
+        try frc.performFetch()
+        // Set initial state on the calling thread, which requires to use performAndWait
+        queue.async { [frc, context, itemCreator, sorting, weak self] in
+            var collection: StreamCollection<Item>!
+            context.performAndWait {
+                collection = Self.makeCollection(frc: frc, context: context, itemCreator: itemCreator, sorting: sorting)
+            }
+            self?._items = collection
+        }
```

Comment:
On the second look, this is not strictly required every single time so I should make it configurable.